Pagination for sign/verify message

matejcik opened the pull request:
based on and supersedes #1337 

The biggest part of this PR is actually a refactor of the `render_text` function, which adds three new outward-facing features:

- the `Span` class can be used to count lines in a string
- breaking lines on whitespace (enabled by default)
- rendering embedded newlines

Includes @mmilata's patch to modtrezordisplay that allows rendering substrings, and takes advantage of this in `render_text` to avoid slicing passed-in strings. This causes rather solid savings in memory allocations: [114k](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/909076526/artifacts/core/prof/memperf-html/index.html) allocs in text.py vs master's ~147k (i don't have a link to an artifact with this :( )
notably, there are **zero allocations** in calling `render_text`

---

the newlines rendering means that we can use the following going forward
old:
```
text.normal("Do you really", "want to change", "your PIN?")
```
new:
```
text.normal("Do you really\nwant to change\nyour PIN?")
```
which is an important step in #1005

mmilata commented on tests/device_tests/test_msg_signmessage.py:
```diff
@@ -206,3 +206,42 @@ def test_signmessage(client, coin_name, path, script_type, address, message, sig
+
+MESSAGE_LENGTHS = (
+    pytest.param("This is a very long message. " * 64, id="normal_text"),
+    pytest.param("ThisIsAMessageWithoutSpaces" * 64, id="no_spaces"),
+    pytest.param("ThisIsAMessageWithLongWords " * 64, id="long_words"),
```
Hmm can you perhaps add another two cases like this but with lots of utf8 continuations?

matejcik replied:
that would be better in a unit test for core, no?

anyway i might need to add unit tests for the `Span` class anyway so I'll keep that in mind

mmilata replied:
Right, I should improve my unit test instead.

mmilata replied:
Feel free to cherry-pick ad05f8ade48eacef18fe238c02792b125c755aa8 which has some more test cases.

matejcik replied:
cherry-picked as 9cf0385b

mmilata reviewed (commented):
Nice work! Your `render_text` is much more readable. UI diff looks legit.

matejcik commented on core/src/apps/management/apply_settings.py:
```diff
@@ -150,8 +150,7 @@ async def require_confirm_change_display_rotation(ctx, rotation):
     text = Text("Change rotation", ui.ICON_CONFIG, new_lines=False)
-    text.normal("Do you really want to", "change display rotation")
-    text.normal("to")
+    text.normal("Do you really want to\nchange display rotation\nto")
```
not sure honestly, I think I needed to switch to linebreaks in some previous iteration of the renderer. I'll retry and remove this line if it turns out that it's not necessary.

matejcik replied:
(fwiw we want to move to single-strings everywhere anyway though, so that we can then extract them for translation)

matejcik replied:
reverted in 6206569

matejcik commented on core/src/apps/common/layout.py:
```diff
@@ -136,3 +136,59 @@ async def show_success(
+
+
+def paginate_text(
+    text: str,
```
No, string formatting is not the way to go here.

My reasoning is that we don't actually need (right now) to smart-paginate anything other than long blocks of text; altcoin property screens should use manual linebreaks and so know in advance how many screens they need.
(in the redesign, it seems that there should only be one property per page anyway?)

mmilata replied:
Even with property-per-page, some of the properties have varying size that may or may not fit on one screen (addresses, arbitrary user inputs). But yeah we don't need pagination of `List[TextContent]` for that, specialized function for paginating property lists probably makes more sense.

mmilata commented on core/src/trezor/ui/text.py:
```diff
@@ -15,109 +15,255 @@
+
+        span.reset(
+            item,
+            char_offset,
```
Shouldn't this apply only to the first item? I.e. something like `char_offset if item_index == item_offset else 0`?

matejcik replied:
fixed in 0e13f417

mmilata commented on core/src/trezor/ui/text.py:
```diff
@@ -15,109 +15,255 @@
+    If `break_words` is false (default), linebreaks will only be rendered (a) at
+    whitespace, or (b) in case a word does not fit on a single line. If true, whitespace
+    is ignored and linebreaks are inserted after the last character that fits.
+    """
```
Ran into this corner case where `\n` gets printed:
![emu00000000](https://user-images.githubusercontent.com/85857/102286324-5ca57d80-3f38-11eb-81ad-4a9f4394a2b7.png)
```
text = Text("title", new_lines=False)
text.normal("1234567890 1234567", "\nfoo")
return interact(ctx, Confirm(text), "test")
```
Was trying to write naive reimplementation of `format_parametrized` and this sometimes happened with `foo\n{}\nbar` format.

matejcik replied:
nice catch! this is because of the `XXX hack` part which tries to render the string as a whole if it fits on a line by itself.

Wondering what to do about it. I want to get rid of the whole branch, but that introduces too many UI diffs at the moment that would need to be fixed manually.

OTOH if we wait with merging this until we have the redesign ready, we can do exactly that: remove the hack branch and make sure no layouts are relying on that behavior.

Is it a problem for you to patch out the hack branch and continue working on that version? Or do you also rely on the good UI results?
In any case this *seems* fixable by adding `and "\n" not in item` to the hack condition?

mmilata replied:
Currently I'm aiming for as little UI changes as possible, however removing the branch after redesign is ready works for me, will work around it in the meantime.

mmilata replied:
Can we just modify the branch to do a newline and then go on with rendering as usual? PTAL @ 3d375e86 Should produce no [UI diff](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/920002077).

matejcik replied:
yes, this seems reasonable. I'll cherry-pick that commit

matejcik replied:
OTOH it seems that in your case it will cause an additional newline, right? that's not good.

mmilata replied:
Yes, good point, didn't realize this and no UI-recorded test triggers it. Would it be too much of a hack to detect the case and eat the `\n`?
```python
        item_width = ui.display.text_width(item, font)
        if (
            item_width <= line_width
            and item_width + offset_x - INITIAL_OFFSET_X > line_width
        ):
            offset_x = INITIAL_OFFSET_X
            offset_y += TEXT_LINE_HEIGHT
            if len(item) > char_offset and item[char_offset] == "\n":
                char_offset += 1

```

matejcik replied:
I'd prefer if we avoided hitting the hack branch at all in new code. We don't want to create layouts that rely on its existence.
Adding `and "\n" not in item` to the hack condition will fully skip it for new-style strings, which I believe is acceptable.

mmilata replied:
Works for me!

matejcik replied:
resolved in 82ce9be47460baa8a3f4857f3f9ff9170a1ce95c
sorry for not cherry-picking, i had some git mess to clean here and the cherrypick commit disappeared in the process :(

matejcik commented:
added some unit tests in c2e73a1

changed behavior slightly in 49b55b5, this will introduce UI diffs that i won't be fixing at the moment. Feel free to ignore that commit until I update the UI fixtures, or review the diffs and push the necessary changes here if you like.

mmilata commented:
[The diffs](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/915191145/artifacts/test_ui_report/index.html) are OK, pushed fixtures in eebefff.

matejcik commented on core/src/trezor/ui/text.py:
```diff
@@ -128,14 +274,25 @@ def __init__(
+        self.content_offset = content_offset
+        self.char_offset = char_offset
+        self.line_width = line_width
+        self.repaint = True
```
fixed in 82ce9be4

matejcik commented:
this is now ready for review, modulo passing CI checks.

@prusnak you're tagged for reviewing the C rendering code changes

for @mmilata:
the last collection of commits add more unit tests and some docs for `Span`, plus two API functions that are used in the updated `paginate_text` function. It is now more readable and more reliable, and also fixes some corner cases (trailing newlines) that were difficult to express properly before.

I implemented the "mock out `ui.display`" way of unit-testing, the final version is as clean as I believe possible for doing this particular sort of thing :)

There is also a proper UI test for the paginated signmessages

matejcik commented:
[UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/927348875/artifacts/master_diff/index.html)

mmilata commented on tests/device_tests/test_msg_signmessage.py:
```diff
@@ -206,3 +206,79 @@ def test_signmessage(client, coin_name, path, script_type, address, message, sig
+    message_read = ""
+    message += "End."
+
+    def input_flow():
```
Out of scope for this PR but I'm wondering how to generalize this for other tests to use. Would be nice to record fixtures for all the pageable screens in tests that have them.

matejcik replied:
indeed, we'd need a debuglink indication that a screen is paginated, and ideally a page position as well.

relatively easy to implement, but i'm thinking maybe the redesign of ButtonRequests will solve this too?

mmilata reviewed (approved):
LGTM!!! Left some comments inline.

Went through the UI diff, the changes are either barely perceptible or a clear improvement to the previous state.

matejcik commented on core/src/trezor/ui/text.py:
```diff
@@ -15,109 +15,336 @@
+    the first character of the indicated item which should be considered.
+    The purpose is to allow rendering different "pages" of text, using the same `items`
+    argument (slicing the list could be expensive in terms of memory).
+    The item selected by `item_offset` must be a string.
```
This is kind of a problem actually, it reveals a bug: font instructions that precede `item_offset` are going to be ignored.

matejcik replied:
fixed in 02ca00d

matejcik commented on core/src/trezor/ui/text.py:
```diff
@@ -15,109 +15,336 @@
+    def next_line(self) -> bool:
+        """Advance the span to point to contents of the next line.
+
+        Returns True if the rendered should make newline afterwards, False if this is
```
fixed in a83f110

matejcik commented on core/src/trezor/ui/text.py:
```diff
@@ -15,109 +15,336 @@
+            if item is BR or item is BR_HALF:
                 # line break or half-line break
                 if offset_y > offset_y_max:
                     ui.display.text(offset_x, offset_y, "...", ui.BOLD, ui.GREY, bg)
```
fixed in a83f110

matejcik commented:
rebased on master, added changelogs.

ping @prusnak for C review, otherwise ready to merge

prusnak commented on core/embed/extmod/modtrezorui/display.c:
```diff
@@ -867,3 +867,38 @@ void display_fade(int start, int end, int delay) {
+
+#define UTF8_IS_CONT(ch) (((ch)&0xC0) == 0x80)
+
+void display_utf8_substr(char *buf_start, size_t buf_len, int char_off,
+                         int char_len, char **out_start, int *out_len) {
```
```suggestion
void display_utf8_substr(const char *buf_start, size_t buf_len, int char_off,
                         int char_len, const char **out_start, int *out_len) {
```

prusnak commented on core/embed/extmod/modtrezorui/display.h:
```diff
@@ -113,4 +113,8 @@ int display_orientation(int degrees);
 void display_fade(int start, int end, int delay);
 
+// helper for locating a substring in buffer with utf-8 string
+void display_utf8_substr(char *buf_start, size_t buf_len, int char_off,
+                         int char_len, char **out_start, int *out_len);
```
```suggestion
void display_utf8_substr(const char *buf_start, size_t buf_len, int char_off,
                         int char_len, const char **out_start, int *out_len);
```

prusnak commented on core/embed/extmod/modtrezorui/modtrezorui-display.h:
```diff
@@ -331,10 +336,25 @@ STATIC mp_obj_t mod_trezorui_Display_text(size_t n_args, const mp_obj_t *args) {
   mp_int_t bgcolor = mp_obj_get_int(args[6]);
-  display_text(x, y, text.buf, text.len, font, fgcolor, bgcolor);
+
+  char *buf_start = text.buf;
```
```suggestion
  const char *buf_start = text.buf;
```

prusnak commented on core/embed/extmod/modtrezorui/modtrezorui-display.h:
```diff
@@ -397,20 +417,45 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorui_Display_text_right_obj,
+  mp_get_buffer_raise(args[1], &txt, MP_BUFFER_READ);
+  mp_int_t f = mp_obj_get_int(args[2]);
+
+  char *buf_start = txt.buf;
```
```suggestion
  const char *buf_start = txt.buf;
```

prusnak reviewed (changes requested):
Let's use `const char *` instead of `char *` to indicate we are not modifying the contents of the string, just slicing it.

mmilata commented:
Feel free to cherry-pick 1c9e24754d630bd5eb76fcad5f025434f9f7fea2 with the suggested changes.

matejcik commented:
cherry-pick applied
